### PR TITLE
Add timeout for `istioctl analyze` getting resources

### PR DESCRIPTION
**Pull request by sushicw**
Add a timeout when pulling resources from the Kubernetes api server, primarily to address https://github.com/istio/istio/issues/18177. The previous behavior was to keep trying indefinitely, even if it would never succeed (e.g. because permissions were missing).

Unfortunately I couldn't find a hook into the k8s go client to detect these errors specifically and handle them more gracefully (e.g. less log spam, fail immediately instead of timing out). This PR at least lets `analyze` terminate, even if it does generate a boatload of error messages in the process. 

*EDIT: We can do a permissions pre-check as a separate PR, but having a timeout on `istioctl analyze` is a good idea anyway. This PR still covers that, ~~as well as adding a role that encapsulates the permissions needed to do analysis. (Reasoning for this [here](https://github.com/istio/istio/pull/19890#discussion_r362659871).)~~*

### CODE REVIEW

File: install/kubernetes/helm/istio/templates/clusterrole.yaml
Comment (howardjohn) [inline install/kubernetes/helm/istio/templates/clusterrole.yaml]: doesn't galley already have all of this? I am not sure why we need to add this

Comment (howardjohn) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: I see from the doc maybe it is useful, but we probably don't want the role binding?

Comment (sushicw) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: These are the goals I'm aiming at, and I'm very open to ideas on how to more elegantly hit them.

1. Have an Istio-maintained role so that it's straightforward to bind a user or service account that can run `istioctl analyze` without permission errors.
2. Have a canonical location documenting exactly what permissions `analyze` needs. (It's very close to what Galley uses, but not quite identical). 
3. Make sure that as we add analyzers that may depend on new  resources, Galley has the right permissions to run them when using `enableAnalysis`. For example, we added an analyzer that looks for valid references to k8s `secret` resources, but forgot to update Galley's permissions to include being able to read that. Having one role that's canonical for all analysis cases should make it easier to avoid that kind of mistake.

Comment (jasonwzm) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: I am still not understanding why the clusterrolebinding is needed as a helm template. Can we just mention in the docs that in cases where they need granular permissions, use the cluster role and bind to account they need to use for analyze?

Comment (sushicw) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: The binding is for the Galley status update case, so that the permissions used for analyze by Galley are consistent with the permissions used for analyze by istioctl. 

Alternatives I considered, but like less, are to either maintain separate roles for istioctl and galley (and update both when analyzers need new permissions), or to expect users to bind to the Galley role for the purpose of using istioctl.

Comment (jasonwzm) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: discussed offline. LGTM now.

**Review by jasonwzm — APPROVED** [review]

File: install/kubernetes/helm/istio/templates/clusterrole.yaml
Comment (sushicw) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: @howardjohn Does this reasoning make sense to you? Happy to discuss offline if it would help. If you agree this approach is OK, I need an env WG approver for this PR.

**Comment by sushicw** [comment]
Holding this, after talking with ayj@ I want to investigate a couple lines of possibility:

1. Doing a permissions check up-front with the `SelfSubjectAccessReview` API (see [here](https://kubernetes.io/docs/reference/access-authn-authz/authorization/#checking-api-access)).
2. Pushing the timeout up to be a general timeout around analyze. (Even if we more elegantly solve the permissions issue, we still want to have a timeout bound).

**Comment by sushicw** [comment]
Updated to make the timeout operate at the entire local analyzer level instead of being per-watcher. This is a more general timeout for the command that covers everything, configurable on the command line.

I'm going to handle doing a permissions pre-check as a separate PR.

**Review by selmanj — APPROVED** [review]

**Review by elfinhe — APPROVED** [review]
LGTM

File: install/kubernetes/helm/istio/templates/clusterrole.yaml
Comment (ayj) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: What's the delta between the existing galley role and this new one?

Comment (sushicw) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: @ayj The Galley role has "get" in addition to "list" and "watch", but I'm not sure it even uses them.

The galley role additionally has some "update" permissions for `validatingwebhookconfigurations` and `deployments/finalizers`. And "update" permissions on the status subresource for Istio CRDs.

The galley role does *not* have permissions to read the `configmap` resource for mesh config. It gets that data via the file mount instead. It's also missing list/watch permissions for the `secrets` resource, which is used by an analyzer.

Comment (sushicw) [reply install/kubernetes/helm/istio/templates/clusterrole.yaml]: Talked with @ayj some more and decided we can hold off on adding a special clusterrole or clusterrolebinding for now. The future maintenance burden of adding something new here is significant, and we don't know that it's even in demand. When permissions are missing, the error messages are very clear, and I expect it wouldn't take someone long to figure out what RBAC their user needs if they're running as a nonprivileged user. Given that, adding a maintenance burden in order to save a relatively small amount of user work is probably premature.

**Review by ayj — APPROVED** [review]

**Pull request merged**
